### ISSUE TITLE
VImage to QImage

### ISSUE THREAD

**Issue by johndoe46** [jcupitt/libvips#1060]
Hello,

I'm not sure where to ask for help so I'll open an issue here.

I used pyvips and pyside2 to open a very big picture (11GB tiled Tiff), crop it and return a smaller QImage. In Python it works great and is easy to write, so I now have a module that performs all kinds of conversions between Vips, QT and Numpy (I let it below for reference). Unfortunately, I have to convert part of this module to C++ and I can't find the same API between pyvips and vips-cpp:

Given a ```pyvips.Image```, there's a ```crop()``` method. But there is no such method on a ```VImage```. There's an ```extract()``` method but it seems different, loading the big picture in memory.

How can I achieve something similar to this in C++ ?

```python
import pyvips
# this is fast
big_pic =  pyvips.Image.new_from_file("/path/to/big.tif")
small_pic = big_pic.crop(top, left, width, height)
```

So far, this is what I do:

```cpp
VImage big_pic = VImage::new_from_file(panoramaPath.toLatin1().data()); // yes... ugly Qstring conversion
VImage small_pic = in.extract_area(top, left, width, height);
```

But it's (very) slow and I suspect it's reading the entire big picture.

Thank you for your help.

(Python module I use, may help some readers)

```python
import pyvips
import numpy
from PySide2.QtGui import QImage, QPixmap
import qimage2ndarray

dtype_to_format = {
    "uint8": "uchar",
    "int8": "char",
    "uint16": "ushort",
    "int16": "short",
    "uint32": "uint",
    "int32": "int",
    "float32": "float",
    "float64": "double",
    "complex64": "complex",
    "complex128": "dpcomplex",
}

format_to_dtype = {
    "uchar": numpy.uint8,
    "char": numpy.int8,
    "ushort": numpy.uint16,
    "short": numpy.int16,
    "uint": numpy.uint32,
    "int": numpy.int32,
    "float": numpy.float32,
    "double": numpy.float64,
    "complex": numpy.complex64,
    "dpcomplex": numpy.complex128,
}


def numpy_to_vips(numpy_3d: numpy.ndarray) -> pyvips.Image:
    height, width, bands = numpy_3d.shape
    linear = numpy_3d.reshape(width * height * bands)
    return pyvips.Image.new_from_memory(
        linear.data, width, height, bands, dtype_to_format[str(numpy_3d.dtype)]
    )


def vips_to_numpy(vips_image: pyvips.Image) -> numpy.ndarray:
    return numpy.ndarray(
        buffer=vips_image.write_to_memory(),
        dtype=format_to_dtype[vips_image.format],
        shape=[vips_image.height, vips_image.width, vips_image.bands],
    )


def numpy_to_qimage(numpy_3d: numpy.ndarray) -> QImage:
    return qimage2ndarray.array2qimage(numpy_3d)


def vips_to_qimage(vips_image: pyvips.Image) -> QImage:
    return numpy_to_qimage(vips_to_numpy(vips_image))


def path_to_vips(path: str) -> pyvips.Image:
    return pyvips.Image.new_from_file(path)


def path_to_qimage(path: str, crop: tuple = None) -> QImage:
    if not crop:
        return QImage(path)
    vips_im = path_to_vips(path)
    if crop:
        left, top, width, height = crop
        vips_im = vips_im.crop(left, top, width, height)
    return vips_to_qimage(vips_im)


def path_to_qpixmap(path: str, crop: tuple = None) -> QPixmap:
    return QPixmap(path_to_qimage(path, crop))

**Comment by johndoe46** [comment]
from the docs:

> A synonym for vips_extract_area(). 

So I must be doing something wrong.

I'll close this for now.

**Comment by jcupitt** [comment]
You can set `access="sequential"` in `new_from_file`. That'll stop the whole image being loaded on open. 

Tiled TIFFs are very fast to crop, since they support true random access.

**Comment by johndoe46** [comment]
It turns out I was using a tiled tiff for my Python tests and a non-tiled tiff for my cpp tests... stupid me. Now the cpp version is as good as the py version (!).

Thank you for your help and most of all for the incredible library that is Vips.

**Comment by jcupitt** [comment]
I wrote a SO answer with some timings for crop:

https://stackoverflow.com/a/48376269/894763

tldr: you can get faster cropping by setting `access`.

**Comment by johndoe46** [comment]
Great, good to know because I have to handle a 483753x5550 px tiff now :-)

Awesome work with the Python bindings too, thank you.

**Comment by johndoe46** [comment]
Yes it is ! In fact it's a 11 gigabytes pyramidal tiled tiff:

```
➜  identify /home/johndoe46/pics/big.tiled.tif
/home/johndoe46/pics/big.tiled.tif[0] TIFF64 483753x5550 483753x5550+0+0 8-bit sRGB 10.988GB 0.010u 0:00.000
/home/johndoe46/pics/big.tiled.tif[1] TIFF64 241876x2775 241876x2775+0+0 8-bit sRGB 10.988GB 0.010u 0:00.000
/home/johndoe46/pics/big.tiled.tif[2] TIFF64 120938x1387 120938x1387+0+0 8-bit sRGB 10.988GB 0.010u 0:00.000
/home/johndoe46/pics/big.tiled.tif[3] TIFF64 60469x693 60469x693+0+0 8-bit sRGB 10.988GB 0.010u 0:00.000
/home/johndoe46/pics/big.tiled.tif[4] TIFF64 30234x346 30234x346+0+0 8-bit sRGB 10.988GB 0.000u 0:00.000
/home/johndoe46/pics/big.tiled.tif[5] TIFF64 15117x173 15117x173+0+0 8-bit sRGB 10.988GB 0.000u 0:00.000
/home/johndoe46/pics/big.tiled.tif[6] TIFF64 7558x86 7558x86+0+0 8-bit sRGB 10.988GB 0.000u 0:00.000
/home/johndoe46/pics/big.tiled.tif[7] TIFF64 3779x43 3779x43+0+0 8-bit sRGB 10.988GB 0.000u 0:00.000
/home/johndoe46/pics/big.tiled.tif[8] TIFF64 1889x21 1889x21+0+0 8-bit sRGB 10.988GB 0.000u 0:00.000
/home/johndoe46/pics/big.tiled.tif[9] TIFF64 944x10 944x10+0+0 8-bit sRGB 10.988GB 0.000u 0:00.000
/home/johndoe46/pics/big.tiled.tif[10] TIFF64 472x5 472x5+0+0 8-bit sRGB 10.988GB 0.000u 0:00.000
/home/johndoe46/pics/big.tiled.tif[11] TIFF64 236x2 236x2+0+0 8-bit sRGB 10.988GB 0.000u 0:00.000
```
Sequential access seems to handle this just fine :)

**Comment by jcupitt** [comment]
Heh well that's good. 

Seq access won't make any difference for tiled tiff: it's just always fast.

**Comment by Topy44** [comment]
> (Python module I use, may help some readers)

Actually, that does exactly what I need, thank you. Under what license can I use it? My project will most likely be GPL.

**Comment by johndoe46** [comment]
> > (Python module I use, may help some readers)
> 
> Actually, that does exactly what I need, thank you. Under what license can I use it? My project will most likely be GPL.

well... I never thought about it... so I'll go with the WTFPL (see https://en.wikipedia.org/wiki/WTFPL). Hopefully this should accomodate your needs. Should you need another license, just let me know and I'll happily comply to whatever you need.

**Comment by erhuodaosi** [comment]
@jcupitt @johndoe46 
Hi,

I'm sorry to bother you,but I have some questions about the vipImage to QImage.Given a vipImage,how could we transform it into vipImage in C++?

I'm looking forward to receiving your reply!
I would very appreciate if you could help me,thank you!

Best,
FengLang

**Comment by jcupitt** [comment]
Hello @erhuodaosi,

The API is the same -- you use `write_to_memory` to write a VImage to a memory area ready for QImage, and `new_from_memory` to wrap a VImage around a pointer from QImage.

https://github.com/libvips/libvips/blob/master/cplusplus/include/vips/VImage8.h#L497

https://github.com/libvips/libvips/blob/master/cplusplus/include/vips/VImage8.h#L574

**Comment by jcupitt** [comment]
(better C++ docs are coming in the next version of libvips)

**Comment by erhuodaosi** [comment]
@jcupitt 

Hi,

Thank you for your patient and timely reply!
I would have a try!
Thank you very much!

Best,
FengLang

**Comment by tks-reply** [comment]
@jcupitt 
 
I desire to add a way to warp index in krita using libvips,could you please help me check the code?

Here is the code( // are what I add):

//add libvips
#include <vips/vips.h>

QImage KisWarpTransformStrategy::calculateTransformedImage(ToolTransformArgs &currentArgs,
                                                           const QImage &srcImage,
                                                           const QVector<QPointF> &origPoints,
                                                           const QVector<QPointF> &transfPoints,
                                                           const QPointF &srcOffset,
                                                           QPointF *dstOffset)
{
 //define the index using libvips
   QImage  *indexImage;
   VipsImage **index;
   int width = srcImage.width();
   int height =srcImage.height();
   vips_xyz (VipsImage **index,width,height);
   
 //tansform the index  to QImage type
   *indexImage = QImage((uchar*)**index.data(),  **index.width(), **index.height(), VIPS_IMAGE_SIZEOF_LINE( **index.get_image()), QImage::Format_RGB888);

 //transformed index
    QImage  t;
     t = KisWarpTransformWorker::transformQImage(
        currentArgs.warpType(),
        origPoints, transfPoints,
        currentArgs.alpha(),
        indexImage,
        srcOffset, dstOffset);

//transform the QImage into index  and save the index to disk
    t.save("transformed_index.v","V");
        
    return KisWarpTransformWorker::transformQImage(
        currentArgs.warpType(),
        origPoints, transfPoints,
        currentArgs.alpha(),
        srcImage,
        srcOffset, dstOffset);
}

I would very appreciate if you could help me,thank you very much!

**Comment by jcupitt** [comment]
Hi @tks-reply,

You need to include `<vips/vips8>` to get the C++ API, and it's called `VImage`, not `VipsImage`. 

You seem to just be using `xyz` -- I doubt if this will be any better than the Krita functions. You'd probably need to compute the warp using libvips as well to see an improvement.

There are lots of ways of computing the warp. I think you'd need to read up on techniques there first.

**Comment by tks-reply** [comment]
@jcupitt

Thank you for your patient.
I'm sorry to bother you,but due to my poor programming in C++ and libvips，I still have some questions.
If I just want to use krita's method to deform the index picture,how do I do that?
1.In python ,we can use  "index = pyvips.Image.xyz(width, height)" to get an index picture.
In C, we can use vips_xyz (VipsImage **out, int width, int height,...); to create an index picture.
But how could we get an index picture in C++?
2.there is something wrong with the transformation between VImage and QImage.

//tansform the VImage to QImage type
indexImage = QImage((uchar)index.data(), width, height, VIPS_IMAGE_SIZEOF_LINE( index.get_image()), QImage::Format_RGB888);

I would very appreciate if you could help me,thank you very much!

**Comment by jcupitt** [comment]
1. Yes, that's correct.
2. It looks OK at a quick glance. What error do you get?
3. Sorry, I don't know QImage.

**Comment by tks-reply** [comment]
@jcupitt 

Thank you for your  patient and timely reply!

The error is listed as below:
1.error: no matching function for call to ‘QImage(const void*, int&, int&, guint64, QImage::Format)’
2. error: invalid conversion from ‘const void*’ to ‘const uchar* {aka const unsigned char*}’ [-fpermissive]   (index.data())
3. error: invalid conversion from ‘const void*’ to ‘uchar* {aka unsigned char*}’ [-fpermissive]
(index.data())

I would very appreciate if you could help me,thank you very much!

**Comment by tks-reply** [comment]
@jcupitt 

I have modified the code,the code is listed as below:

 `    vips::VImage index = VImage::xyz( width , height ) ;
    index.cast(VIPS_FORMAT_UCHAR);
    //tansform the index  to QImage type
    QImage  *indexImage = QImage((uchar*)index.data(), width, height, VIPS_IMAGE_SIZEOF_LINE( index.get_image()), QImage::Format_RGB888);`

But I get a new error "error: cannot convert ‘QImage’ to ‘QImage*’ in initialization
 th, height, VIPS_IMAGE_SIZEOF_LINE( index.get_image()), QImage::Format_RGB888);".

How could I solve it?
I would very appreciate if you could help me,thank you very much!

**Comment by tks-reply** [comment]
@jcupitt 
 
I have modified my code as below:   

    ` QImage *indexImage;
    *indexImage = QImage((uchar*)index.data(), width, height, VIPS_IMAGE_SIZEOF_LINE( index.get_image()), QImage::Format_RGB888);`

But it returns some errors:
![2020-08-28 09-37-34 的屏幕截图](https://user-images.githubusercontent.com/59277770/91511416-2d7be980-e912-11ea-8f4e-ca736a2a3710.png)

I have rebuilt the libvips but in vain.
How could I solve it?

I would very appreciate if you could help me,thank you very much!

**Comment by jcupitt** [comment]
```
undefined reference to xyz,cast,vips_object_get_type,g_type_check_instance_is_a,g_assertion_message_expr,g_object_unref,vips_format_sizeof_unsafe
```

That probably means you are not setting your compiler flags correctly, or you have built libvips with a different C++ compiler from your application.

Have you tried building a tiny "hello world" libvips C++ program? There's a test one here:

https://libvips.github.io/libvips/API/current/using-from-cpp.html

**Comment by tks-reply** [comment]
@jcupitt 

Thank you for your help!
The test could run normally.I think maybe there is something  wrong with glib.

**Comment by jcupitt** [comment]
Do you mean you can compile and run that test program?

If that's working, then the problem is the way you are compiling your krita code. What flags are you giving to the compiler? Compare them to the flags for the test program and see what's different.

**Comment by tks-reply** [comment]
@jcupitt 

Thank you for your patient!
Yes, I could compile and run that test program,and it generates a file called a.out.
I just put "make -j8" to compile my modified krita code,I could run normally without adding  those code, but failed with adding modfied code.It seems that I could not use the function defined in libvips.

![2020-08-28 17-40-06 的屏幕截图](https://user-images.githubusercontent.com/59277770/91546763-8c168700-e955-11ea-8d15-6bd9066d4a88.png)
![2020-08-28 17-40-48 的屏幕截图](https://user-images.githubusercontent.com/59277770/91546813-a2bcde00-e955-11ea-8262-45345916d139.png)

**Comment by jcupitt** [comment]
It looks like Krita is using cmake. You'll need to change the Krita `CMakeLists.txt` to add libvips as a dependency.

**Comment by tks-reply** [comment]
@jcupitt 

Thank you very much!You help me a lot.
It is too embarrassed to say,but I really need your help...
It  is the last step to rebuild krita,I have a strong desire to see the results in krita,really,really...
But I am new to C++、libvips and QImage , I am so weak in them .
How to change the cmakefile?Could you give me some advice?
I have changed the makefiles in krita but in vain, the line 20 to 23 is what I add, is there something that I miss?
![2020-08-29 10-43-59 的屏幕截图](https://user-images.githubusercontent.com/59277770/91626875-61740f00-e9e5-11ea-8373-5e693ad22140.png)
 
I would very appreciate if you could help me,thank you very much!

**Comment by jcupitt** [comment]
Hello again, 

Please don't post screenshots -- they are hard to read and I can't edit them or copy text out. Instead, just copy-paste the text itself. You can enclose text with ``` (three backticks) to mark it as computer output.

You don't need to test for vips and for vips-cpp. Just check for vips-cpp.

Don't use `find_package(vips)`. That will look for a vips.cmake, and vips uses pkg-config.

Look at the docs for `pkg_check_modules()`. You might need to add something to the main Krita flags.

**Comment by tks-reply** [comment]
@jcupitt 

Thank you for your suggestions.I am so sorry to put those screenshots.
I have added something to cmakefile,but it still returns another error.The modified code in cmakefile is listed as below:

find_package(PkgConfig)
pkg_check_modules(VIPS vips-cpp>=8.10 REQUIRED)

find_package(PkgConfig REQUIRED)
project(kritadev)
add_executable(kritadev ${/home/fl/kritadev/krita/plugins/tools/tool_transform2/kis_warp_transform_strategy.cpp})
pkg_search_module(vips REQUIRED vips>=8.10.0)
INCLUDE_DIRECTORIES(${vips_INCLUDE_DIRS})
TARGET_LINK_LIBRARIES(kritadev ${vips_LIBRARIES})

The error is:
-- Configuring done
-- Generating done
-- Build files have been written to: /home/fl/kritadev/build
[  0%] Automatic MOC for target kritaversion
[  0%] Automatic MOC for target kritadev
[  0%] Built target kritadev_autogen
[  0%] Built target kritaversion_autogen
[  0%] Linking CXX executable kritadev
[  1%] Built target kritaversion
[  1%] Automatic MOC for target kritaglobal
[  1%] Automatic MOC for target krita_version
AutoMoc warning
---------------
"SRC:/libs/global/kis_signal_compressor_with_param.cpp"
includes the moc file "kis_signal_compressor_with_param.moc", but does not contain a Q_OBJECT, Q_GADGET, Q_NAMESPACE, Q_NAMESPACE_EXPORT, K_PLUGIN_FACTORY_WITH_JSON, K_PLUGIN_FACTORY or K_PLUGIN_CLASS_WITH_JSON macro.

[  1%] Built target krita_version_autogen
[  1%] Built target kritaglobal_autogen
/usr/lib/gcc/x86_64-linux-gnu/7/../../../x86_64-linux-gnu/Scrt1.o：in the function‘_start’：
(.text+0x20): undefined reference to‘main’
collect2: error: ld returned 1 exit status
CMakeFiles/kritadev.dir/build.make:102: recipe for target 'kritadev' failed
make[2]: *** [kritadev] Error 1
CMakeFiles/Makefile2:6723: recipe for target 'CMakeFiles/kritadev.dir/all' failed
make[1]: *** [CMakeFiles/kritadev.dir/all] Error 2
make[1]: *** ready for the unfinished task
[  1%] Built target krita_version
Makefile:159: recipe for target 'all' failed
make: *** [all] Error 2

I would very appreciate if you could help me,thank you very much!

**Comment by jcupitt** [comment]
My guess is it's now setting the flags for libvips, but not for Qt. You'll need both.

**Comment by tks-reply** [comment]
@jcupitt 

I'm so sorry to bother you again and again.
Thank you very much for your patients and suggestions.

**Comment by tks-reply** [comment]
@jcupitt 

I'm sorry to bother you again.
What can we do if we want to transform a srgb picture to a multiband VImage?

For example,the shape of  ```transformed_image.jpg``` is (1600,1200,3),we want to get a VImage called ```index.v``` which has the shape of (1600,1200,2).

I have tried two ways but get wrong results:

```
QImage t = KisWarpTransformWorker::transformIndexImage(
        currentArgs.warpType(),
        origPoints, transfPoints,
        currentArgs.alpha(),
        IndexImage,
        srcOffset, dstOffset);
 t.save("transformed_image.jpg");
VImage Img = VImage::new_from_file("transformed_image.jpg",VImage::option ()->set ("access", VIPS_ACCESS_SEQUENTIAL));
 Img.VImage::write_to_file("index.v");
```

```
QImage t = KisWarpTransformWorker::transformIndexImage(
        currentArgs.warpType(),
        origPoints, transfPoints,
        currentArgs.alpha(),
        IndexImage,
        srcOffset, dstOffset);
 t.save("transformed_image.jpg");
 VImage Img = VImage::new_from_memory(t.data_ptr(),3840000,1600,1200,2, VIPS_FORMAT_UCHAR);
 Img.write_to_file("index.v");
```

When we use  ``` vipsheader -a index.v``` in terminal:
The true return should be:
```
index.v: 1086x612 float, 2 bands, multiband
width: 1600
height: 1200
bands: 2
format: 6 - float
coding: 0 - none
interpretation: 0 - multiband
xoffset: 0
yoffset: 0
xres: 1.000000
yres: 1.000000
filename: "index.v"
xml-header: VIPS_TYPE_AREA, count = 1, data = 0x5585b652f600
```

But I get:
```
index.v: 1600x1200 uchar, 3 bands, srgb, jpegload
width: 1600
height: 1200
bands: 3
format: 0 - uchar
coding: 0 - none
interpretation: 22 - srgb
xoffset: 0
yoffset: 0
xres: 3.779528
yres: 3.779528
filename: "index.v"
xml-header: VIPS_TYPE_AREA, count = 1, data = 0x563876734600
vips-loader: jpegload
jpeg-multiscan: 0
```

Could you give me some advice?
I would very appreciate if you could help me,thank you very much!

**Comment by tks-reply** [comment]
@jcupitt 

I have read the libvips documentation and relevant issues seriously. But I just do some research in python rather than C++.
To be honest, I nearly used C++,I'm really weak in it.

I have no chance to resume my work,your suggestions may help me save my job.
Looking forward to receiving your reply.
I would very appreciate if you could help me,thank you very much!

**Comment by jcupitt** [comment]
Hi again, I'm not sure I understand what you are doing, but I made you a demo program:

```C++
/* compile with:
 *
 * g++ xyz.cpp `pkg-config vips-cpp --cflags --libs`
 */

#include <vips/vips8>

using namespace vips;

int
main( int argc, char **argv )
{       
        if( VIPS_INIT( argv[0] ) )
                vips_error_exit( NULL );

        // read the image header
        VImage im = VImage::new_from_file( argv[1], VImage::option ()
                ->set( "access", VIPS_ACCESS_SEQUENTIAL ) );

        // make an index image of that size
        VImage index = VImage::xyz( im.width(), im.height() );
        
        // save the index image to disc
        index.write_to_file( argv[2] );

        return( 0 );
}       
```

I can run it like this:

```
john@yingna ~/try $ g++ xyz.cpp `pkg-config vips-cpp --cflags --libs`
john@yingna ~/try $ ./a.out ~/pics/k2.jpg index.v
john@yingna ~/try $ vipsheader index.v
index.v: 1450x2048 uint, 2 bands, multiband
```

**Comment by tks-reply** [comment]
@jcupitt 

Thank you very very very much for your patient.
Please forgive my poor English.

What I want to do is：

//trans.jpg: 1600x1200 uchar, 1 band, b-w, jpegload
1.Define a VImage,and save it into the disk;

//beauty.jpg: 1600x1200 uchar, 3 bands, srgb, jpegload
2.Load the VImage from the disk,and transform it into QImage;

//transformed_image.jpg: 1600x1200 uchar, 3 bands, srgb, jpegload
3.Transform the QImage using the krita function,and save it into the disk;

//index.v: 1600x1200 uchar, 3 bands, srgb, jpegload
4.Load the transformed QImage and transfrom the QImage to VImage,and save the VImage as the format of  "index.v".

When I do the last step,I want to get a index.v which is  ``` 1600x1200 uint, 2 bands, multiband``` ,but what I really get is   ```1600x1200 uchar, 3 bands, srgb, jpegload```,it seems that it would loss some data  during the transformation from  3 bands to 2 bans.

Here is my code:

```
QImage KisWarpTransformStrategy::calculateTransformedIndexImage(ToolTransformArgs &currentArgs,
                                                           QImage &IndexImage,
                                                           const QVector<QPointF> &origPoints,
                                                           const QVector<QPointF> &transfPoints,
                                                           const QPointF &srcOffset,
                                                           QPointF *dstOffset)
{
    vips::VImage index = VImage::xyz( 1600 , 1200 ) ;
    index.write_to_file("trans.jpg");
    index.write_to_file("indexed.v");

    QPixmap pixmap;
    pixmap.load("trans.jpg");
    IndexImage = pixmap.toImage();
    IndexImage.save("beauty.jpg");
    QImage t = KisWarpTransformWorker::transformIndexImage(
        currentArgs.warpType(),
        origPoints, transfPoints,
        currentArgs.alpha(),
        IndexImage,
        srcOffset, dstOffset);
    t.save("transformed_image.jpg");

    VImage Img = VImage::new_from_file("transformed_image.jpg",VImage::option ()->set ("access", VIPS_ACCESS_SEQUENTIAL));
    Img.VImage::write_to_file("index.v");

    // VImage Img = VImage::new_from_file("transformed_image.jpg",VImage::option ()->set ("access", VIPS_ACCESS_SEQUENTIAL));
    // Img.write_to_file("AA.jpg");
    // Img = VImage::new_from_file("AA.jpg",VImage::option ()->set ("interpretation", VIPS_INTERPRETATION_XYZ));
    // Img.VImage::write_to_file("index.v");

    // VImage Img = VImage::new_from_file("transformed_image.jpg",VImage::option ()->set ("access", VIPS_ACCESS_SEQUENTIAL));
    // Img = VImage::new_from_memory(Img.write_to_memory( ),3840000,1600,1200,2,VIPS_FORMAT_FLOAT);
    // Img.write_to_file("index.v");

    // VImage Img = VImage::new_from_memory(t.data_ptr(),3840000,1600,1200,2, VIPS_FORMAT_FLOAT);
    // Img.write_to_file("index.v");

    return KisWarpTransformWorker::transformIndexImage(
        currentArgs.warpType(),
        origPoints, transfPoints,
        currentArgs.alpha(),
        IndexImage,        
        srcOffset, dstOffset);
 }
```

Could you give me some advice?I have been confused for three days.
Looking forward to receiving your reply.
I would very appreciate if you could help me,thank you very much!

**Comment by jcupitt** [comment]
Sorry, I missed this.

I still don't understand what you are doing :-( I can't find docs for `transformIndexImage` but I'd guess it's an old name for
